### PR TITLE
feat: update rewards history limit default

### PR DESCRIPTION
**Pull request by IvayloAndonov**
# Context
Following our dev huddle with the team, we decided to change the default behavior of Stake pool data fetching with defaulting rewardsHistoryLimit to 3 if not provided as an argument. That would be one of the optimizations related to the bad performance of the stake pool’s load test against `mainnet` .

# Important Changes Introduced
- default rewards history limit to 3 if not specified

### CODE REVIEW

File: packages/cardano-services/test/StakePool/StakePoolHttpService.test.ts
Comment (mkazlauskas) [inline packages/cardano-services/test/StakePool/StakePoolHttpService.test.ts:1367]: If `apy` computation is based on this then maybe `0` is not a good default?

**Review by mkazlauskas — CHANGES_REQUESTED** [review]
ae78fd9200a9fdfde62d36ecd93a39888a2dd6f8 should be scoped to `core`

File: packages/cardano-services/test/StakePool/StakePoolHttpService.test.ts
Comment (IvayloAndonov) [reply packages/cardano-services/test/StakePool/StakePoolHttpService.test.ts:1367]: I think you are right, I just inspected that the `apy` computation requires at least 3 epoch rewards, but with `rewardsHistoryLimit: 0` they are 0, an empty list.
So then maybe defaulting to 3 is more appropriate, given the fact `metrics.apy` is changing depending on each epoch reward from the history. @mkazlauskas

File: packages/cardano-services/src/StakePool/DbSyncStakePoolProvider/DbSyncStakePoolProvider.ts
Comment (iccicci) [inline packages/cardano-services/src/StakePool/DbSyncStakePoolProvider/DbSyncStakePoolProvider.ts:176]: This has impact on APY calculation as well, yes.

While speaking about this activity I was thinkin about change this [code snippet](https://github.com/input-output-hk/cardano-js-sdk/blob/90880cf1be445808b2ca39b3bcd4757bef0f9981/packages/cardano-services/src/StakePool/DbSyncStakePoolProvider/DbSyncStakePoolProvider.ts#L219-L224):

From
```
    // Get stake pools rewards cached
    const poolRewards = await this.#cache.get(
      queryCacheKey(StakePoolsSubQuery.REWARDS, orderedResultHashIds, options),
      () => this.#builder.queryPoolRewards(orderedResultHashIds, this.#epochLength, rewardsHistoryLimit),
      UNLIMITED_CACHE_TTL
    );
```
to
```
    // Get stake pools rewards cached
    const poolRewards = await this.#cache.get(
      queryCacheKey(StakePoolsSubQuery.REWARDS, orderedResultHashIds, options),
      () => this.#builder.queryPoolRewards(orderedResultHashIds, this.#epochLength, typeof rewardsHistoryLimit === 'number' ? rewardsHistoryLimit : 0),
      UNLIMITED_CACHE_TTL
    );
```

Comment (iccicci) [reply packages/cardano-services/src/StakePool/DbSyncStakePoolProvider/DbSyncStakePoolProvider.ts:176]: Probably the cache key, needs to be dependent on `rewardsHistoryLimit` as well.

Comment (IvayloAndonov) [reply packages/cardano-services/src/StakePool/DbSyncStakePoolProvider/DbSyncStakePoolProvider.ts:176]: Because `rewardsHistoryLimit` is used across different places within the method - [here](https://github.com/input-output-hk/cardano-js-sdk/blob/e7084dc9fb0d99aaae4c723263c8afadf3d08b3e/packages/cardano-services/src/StakePool/DbSyncStakePoolProvider/DbSyncStakePoolProvider.ts#L223), [here](https://github.com/input-output-hk/cardano-js-sdk/blob/e7084dc9fb0d99aaae4c723263c8afadf3d08b3e/packages/cardano-services/src/StakePool/DbSyncStakePoolProvider/DbSyncStakePoolProvider.ts#L229) and [here](https://github.com/input-output-hk/cardano-js-sdk/blob/e7084dc9fb0d99aaae4c723263c8afadf3d08b3e/packages/cardano-services/src/StakePool/DbSyncStakePoolProvider/DbSyncStakePoolProvider.ts#L237), I decided to default it at the very beginning in the [options obj destruction](https://github.com/input-output-hk/cardano-js-sdk/blob/e7084dc9fb0d99aaae4c723263c8afadf3d08b3e/packages/cardano-services/src/StakePool/DbSyncStakePoolProvider/DbSyncStakePoolProvider.ts#L176).
But if you see an issue with that I can change it to use inline checks and assignments following your example. @iccicci

File: packages/cardano-services/test/StakePool/StakePoolHttpService.test.ts
Comment (IvayloAndonov) [reply packages/cardano-services/test/StakePool/StakePoolHttpService.test.ts:1367]: Please like if you agree on 3 as the default value for the rewards history limit.

File: packages/cardano-services/src/StakePool/DbSyncStakePoolProvider/DbSyncStakePoolProvider.ts
Comment (iccicci) [reply packages/cardano-services/src/StakePool/DbSyncStakePoolProvider/DbSyncStakePoolProvider.ts:176]: No @IvayloAndonov , I see no issues; mine was just the proposal for another approach (which actually I have not evaluated in the deeps). I'm fine with setting the default value to 3.

**Comment by IvayloAndonov** [comment]
The default is updated to 3 epochs back, commits are squashed.

File: packages/cardano-services/test/StakePool/StakePoolHttpService.test.ts
Comment (rhyslbw) [inline packages/cardano-services/test/StakePool/StakePoolHttpService.test.ts:481]:
```suggestion
          expect(resWithoutRewardsHistoryLimit.pageResults[0].epochRewards).toHaveLength(5);
```
This makes the test more expressive. Same feedback applies for the following test too

File: packages/cardano-services/test/StakePool/StakePoolHttpService.test.ts
Comment (rhyslbw) [inline packages/cardano-services/test/StakePool/StakePoolHttpService.test.ts:479]:
```suggestion
          const resWithoutRewardsHistoryLimit = await provider.queryStakePools(reqWithoutRewardsHistoryLimit);
```

File: packages/cardano-services/src/StakePool/DbSyncStakePoolProvider/util.ts
Comment (rhyslbw) [inline packages/cardano-services/src/StakePool/DbSyncStakePoolProvider/util.ts:11]: Depending on which PR is merged first, let's make this a parameter and use the config to set the default limit to `0` should APY metric be disabled. See @iccicci's comment here: https://github.com/input-output-hk/cardano-js-sdk/pull/609#issuecomment-1435940833

**Review by rhyslbw — CHANGES_REQUESTED** [review]

File: packages/cardano-services/test/StakePool/StakePoolHttpService.test.ts
Comment (IvayloAndonov) [reply packages/cardano-services/test/StakePool/StakePoolHttpService.test.ts:481]: You mean `resWithRewardsHistoryLimit` right? Because this line asserts that we have a rewards history limit of 5.

File: packages/cardano-services/src/StakePool/DbSyncStakePoolProvider/util.ts
Comment (rhyslbw) [reply packages/cardano-services/src/StakePool/DbSyncStakePoolProvider/util.ts:11]: #609 is now on hold, so please ignore my previous comment, and we can handle this change later if needed

File: packages/cardano-services/test/StakePool/StakePoolHttpService.test.ts
Comment (IvayloAndonov) [reply packages/cardano-services/test/StakePool/StakePoolHttpService.test.ts]: 2a42e3b763b44d9f23a0fe9ebf4e47908241b815 @rhyslbw

File: packages/cardano-services/test/StakePool/StakePoolHttpService.test.ts
Comment (IvayloAndonov) [reply packages/cardano-services/test/StakePool/StakePoolHttpService.test.ts]: 2a42e3b763b44d9f23a0fe9ebf4e47908241b815

**Review by rhyslbw — APPROVED** [review]

**Review by rhyslbw — APPROVED** [review]
Nice work @IvayloAndonov :tada:

**Review by iccicci — DISMISSED** [review]
Well done!

**Review by iccicci — DISMISSED** [review]

**Review by rhyslbw — DISMISSED** [review]

File: packages/cardano-services/test/StakePool/StakePoolHttpService.test.ts
Comment (IvayloAndonov) [reply packages/cardano-services/test/StakePool/StakePoolHttpService.test.ts:469]: de6c62ba88c507f3f387766da5c42a8c16792e68 @iadmytro

**Review by iadmytro — DISMISSED** [review]

**Review by iadmytro — APPROVED** [review]

**Review by iccicci — APPROVED** [review]

**Comment by IvayloAndonov** [comment]
Rebased with the master, the conflicts are resolved. Once the CI is passed we can merge it. @rhyslbw

**Comment by rhyslbw** [comment]
Lint error @IvayloAndonov

**Review by iccicci — APPROVED** [review]

**Review by rhyslbw — APPROVED** [review]

**Pull request merged**
